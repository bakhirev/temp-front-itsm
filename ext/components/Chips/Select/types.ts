export interface IItemChipSelect {
  /** Уникальный идентификатор чипса */
  id: string;
  /** Подпись к чипсу */
  label: string;
  /** Отключение чипса */
  disabled?: boolean;
  /** Иконка */
  icon?: any;
}
