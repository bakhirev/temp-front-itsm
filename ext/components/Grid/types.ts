import { ReactNode } from 'react';

/** Вид устройства/разрешения для задания свойств */
export type Device = 'mobile320' | 'tablet768' | 'tablet1024' | 'desktop1280' | 'desktop1600';

export type JustifyContent =
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'space-around'
  | 'space-between';
export type AlignItems = 'baseline' | 'center' | 'flex-start' | 'flex-end' | 'stretch';

export const Devices: Device[] = [
  'mobile320',
  'tablet768',
  'tablet1024',
  'desktop1280',
  'desktop1600',
];

export interface IGridBaseProps {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

/** Содержит размеры в количестве колонок, которое занимает данная ячейка на заданном устройстве/разрешении */
export interface IDevices {
  mobile320?: number;
  tablet768?: number;
  tablet1024?: number;
  desktop1280?: number;
  desktop1600?: number;
}

export interface IGridCellProps extends IGridBaseProps, IDevices {}

export interface IGridProps extends IGridBaseProps {
  /** Отключение отступов внутри модульной сетки */
  noindent?: boolean | undefined;
  /** Отключение отступов снаружи модульной сетки */
  nomargin?: boolean | undefined;
  /** Определяет распределение пространства у элементов контента вдоль главной оси flex контейнера */
  justifyContent?: JustifyContent;
  /** Выравнивает элементы текущей flex-линии аналогично justifyContent, но в перпендикулярном направлении */
  alignItems?: AlignItems;
}

export interface IGridContext {
  noindent?: boolean | undefined;
}
