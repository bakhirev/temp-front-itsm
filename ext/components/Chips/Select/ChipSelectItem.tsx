import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../common';
import { HEIGHT_BIG, HEIGHT_SMALL, MAX_WIDTH, SHAPE_MICRO } from '../common/constants';
import type { Size } from '../Tags/ChipTagItem';

import {
  getChipColor,
  getChipBackgroundColor,
  getChipHoverColor,
  getChipHoverBackgroundColor,
} from './styleUtils';

export interface IChipSelectItemProps {
  size: Size;
  disabled?: boolean;
  selected?: boolean;
  focused?: boolean;
}

export const ChipSelectItem = styled.div<IChipSelectItemProps>`
  ${({ theme, disabled, size, selected }) => css`
    display: flex;
    align-items: center;
    position: relative;
    cursor: ${disabled ? 'default' : 'pointer'};
    padding-left: 8px;
    padding-right: 8px;
    max-width: ${MAX_WIDTH};
    color: ${getChipColor(theme, disabled, selected)};
    background-color: ${getChipBackgroundColor(theme, disabled, selected)};
    border-radius: ${SHAPE_MICRO};
    height: ${size === 'big' ? HEIGHT_BIG : HEIGHT_SMALL};
    ${disabled
      ? ''
      : css`
          &:hover {
            color: ${getChipHoverColor(theme, selected)};
            background-color: ${getChipHoverBackgroundColor(theme, selected)};
          }
        `}
    &:focus {
      outline: none;
      ${disabled ? '' : `border: 2px solid ${theme.color.primary[60]};`}
    }
  `}
`;

ChipSelectItem.defaultProps = {
  theme: DEFAULT_THEME,
};
