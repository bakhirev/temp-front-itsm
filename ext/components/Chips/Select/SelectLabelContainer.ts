import styled, { css } from 'styled-components';

import { TEXT_WIDTH_BIG, TEXT_WIDTH_SMALL } from '../common/constants';
import { Size } from '../common/common';

interface ISelectLabelContainerProps {
  size: Size;
  withIcon: boolean;
  width?: number | string;
}

export const SelectLabelContainer = styled.div<ISelectLabelContainerProps>`
  ${({ size, width, withIcon }) => css`
    display: flex;
    justify-content: center;
    width: ${typeof width === 'number' ? `${width}px` : width};
    max-width: ${size === 'big' ? TEXT_WIDTH_BIG : TEXT_WIDTH_SMALL};
    ${withIcon ? 'margin-left: 6px;' : ''};
  `}
`;
