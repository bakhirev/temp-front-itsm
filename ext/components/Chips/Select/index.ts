export { Select, IChipsSelectProps } from './Select';
export { MultiSelect, IChipsMultiSelectProps } from './MultiSelect';
export { IItemChipSelect } from './types';
