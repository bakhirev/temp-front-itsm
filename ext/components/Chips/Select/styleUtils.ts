export const getChipColor = (theme, disabled, selected) => {
  if (disabled && selected) return theme.color.primary[40];
  if (disabled) return theme.color.neutral[30];
  if (selected) return theme.color.primary[70];

  return theme.color.neutral[90];
};

export const getChipBackgroundColor = (theme, disabled, selected) => {
  if (disabled && selected) return theme.color.primary[20];
  if (disabled) return theme.color.neutral[10];
  if (selected) return theme.color.primary[20];

  return theme.color.neutral[10];
};

export const getChipHoverColor = (theme, selected) => {
  if (selected) return theme.color.primary[70];

  return theme.color.neutral[90];
};

export const getChipHoverBackgroundColor = (theme, selected) => {
  if (selected) return theme.color.primary[30];

  return theme.color.neutral[20];
};

export const getIconColor = (theme, selected, disabled) => {
  if (disabled && selected) return theme.color.primary[40];
  if (disabled) return theme.color.neutral[30];
  if (selected) return theme.color.primary[70];

  return theme.color.neutral[50];
};
