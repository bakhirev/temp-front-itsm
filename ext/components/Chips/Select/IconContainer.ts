import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../common';

import { getIconColor } from './styleUtils';

export interface IIconContainerProps {
  disabled?: boolean;
  selected?: boolean;
}

export const IconContainer = styled.div<IIconContainerProps>`
  ${({ theme, disabled, selected }) => css`
    display: flex;
    justify-content: center;
    & svg {
      fill: ${getIconColor(theme, selected, disabled)};
    }
  `}
`;

IconContainer.defaultProps = {
  theme: DEFAULT_THEME,
};
