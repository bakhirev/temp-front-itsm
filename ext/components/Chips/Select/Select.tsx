import React, { Fragment, FC, useState } from 'react';

import { useTooltip } from '../../common/hooks';
import { createTooltipHOC } from '../../TooltipHOC';
import { KEY_CODES } from '../../common';
import { ICON_SIZE_SMALL, ICON_SIZE_BIG } from '../common/constants';
import { ChipsWrapper } from '../common/ChipsWrapper';
import { Separator } from '../common/Separator';
import type { Size } from '../common/common';
import { StyledCaption11 } from '../common/Caption11';
import { Caption13 } from '../common/Caption13';

import { IItemChipSelect } from './types';
import { ChipSelectItem } from './ChipSelectItem';
import { IconContainer } from './IconContainer';
import { SelectLabelContainer } from './SelectLabelContainer';

export interface IChipsSelectProps {
  /** Массив чипсов */
  items: IItemChipSelect[];
  /** Размер чипсов */
  size?: Size;
  /** Идентификатор выбранного элемента */
  selectedId?: string;
  /** Обработчик выбора элемента */
  onSelectItem?: (id: string) => void;
  /** Обработчик фокусировки на элементе */
  onFocusItem?: (id: string) => void;
  /** В селекте или нет, влияет на размер отступа */
  inSelect?: boolean;
  /** Отключение чипсов */
  disabled?: boolean;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Ширина Тега */
  width?: number | string;
  /** Контейнер, в котором при необходимости можно отрисовать тултип через ReactDOM.createPortal */
  tooltipContainer?: Element | null;
}

export const Select: FC<IChipsSelectProps> = ({
  items,
  inSelect,
  className,
  dataTestId,
  size = 'big',
  selectedId,
  onSelectItem,
  onFocusItem,
  width,
  tooltipContainer,
}) => {
  const [tooltipItems, itemsRef] = useTooltip();
  const [focusedItem, setFocusedItem] = useState(null);
  const [selectedIdState, setSelectedIdState] = useState(
    selectedId || items.length ? items[0].id : null
  );
  const handleFocus = (itemId) => {
    setFocusedItem(itemId);
    onFocusItem && onFocusItem(itemId);
  };
  const handleItemSelect = (itemId) => {
    setSelectedIdState(itemId);
    onSelectItem && onSelectItem(itemId);
  };

  const onKeyDownCreator = (itemId) => (event) => {
    if ([KEY_CODES.SPACEBAR, KEY_CODES.ENTER].includes(event.keyCode)) {
      event.preventDefault();
      handleItemSelect(itemId);
    }
  };

  return (
    <ChipsWrapper className={className} data-test-id={dataTestId}>
      {items.map((chipItem, index) => {
        const Chip = tooltipItems[chipItem.id] ? createTooltipHOC(ChipSelectItem) : ChipSelectItem;
        const Label = size === 'small' ? StyledCaption11 : Caption13;
        const iconSize = size === 'big' ? ICON_SIZE_BIG : ICON_SIZE_SMALL;
        const Icon = chipItem.icon;

        return (
          <Fragment key={chipItem.id}>
            <Chip
              tooltip={chipItem.label}
              container={tooltipContainer}
              size={size}
              disabled={chipItem.disabled}
              onClick={chipItem.disabled ? undefined : () => handleItemSelect(chipItem.id)}
              selected={selectedIdState === chipItem.id}
              onFocus={chipItem.disabled ? undefined : () => handleFocus(chipItem.id)}
              focused={focusedItem === chipItem.id}
              onKeyDown={chipItem.disabled ? undefined : onKeyDownCreator(chipItem.id)}
              tabIndex={chipItem.disabled ? -1 : 1}
            >
              {Icon && (
                <IconContainer
                  disabled={chipItem.disabled}
                  selected={selectedIdState === chipItem.id}
                >
                  <Icon width={iconSize} height={iconSize} />
                </IconContainer>
              )}
              <SelectLabelContainer size={size} width={width} withIcon={!!Icon}>
                <Label ref={(el) => itemsRef(el, chipItem.id)}>{chipItem.label}</Label>
              </SelectLabelContainer>
            </Chip>
            {index !== items.length - 1 && <Separator inSelect={inSelect} />}
          </Fragment>
        );
      })}
    </ChipsWrapper>
  );
};
