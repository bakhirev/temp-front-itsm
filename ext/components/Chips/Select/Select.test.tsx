import React from 'react';
import { shallow } from 'enzyme';

import { IItemChipSelect } from './types';
import { Select } from './Select';

const listData: IItemChipSelect[] = [
  { id: '1', label: 'Суздаль' },
  { id: '2', label: 'Москва', disabled: true },
];

const listItemsDisabled: IItemChipSelect[] = [
  { id: 'a', label: 'Суздаль', disabled: true },
  { id: 'b', label: 'Москва', disabled: true },
  { id: 'c', label: 'Тверь', disabled: true },
  { id: 'd', label: 'Самара', disabled: true },
];

jest.mock('../../common', () => ({
  DEFAULT_THEME: {},
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

describe('Chips.Select', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render component with items', () => {
    const wrapper = shallow(<Select items={listData} size={'big'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with disabled items', () => {
    const wrapper = shallow(<Select items={listItemsDisabled} size={'big'} />);
    expect(wrapper).toMatchSnapshot();
  });
});
