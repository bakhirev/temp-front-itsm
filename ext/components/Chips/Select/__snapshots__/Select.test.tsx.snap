// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips.Select should render component with disabled items 1`] = `
<styled.div>
  <styled.div
    disabled={true}
    focused={false}
    selected={true}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Суздаль"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Суздаль
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Москва"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Москва
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Тверь"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Тверь
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Самара"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Самара
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Chips.Select should render component with items 1`] = `
<styled.div>
  <styled.div
    focused={false}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    selected={true}
    size="big"
    tabIndex={1}
    theme={Object {}}
    tooltip="Суздаль"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Суздаль
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Москва"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Москва
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;
